framework/src/base/NonlinearSystemBase.C

```diff
@@ -1143,27 +1162,35 @@ NonlinearSystemBase::computeNodalBCs(NumericVector<Number> & residual)
   if (_has_save_in)
     _fe_problem.getAuxiliarySystem().solution().close();
 
-  PARALLEL_TRY {
-    // last thing to do are nodal BCs
+  PARALLEL_TRY
+  {
     ConstBndNodeRange & bnd_nodes = *_mesh.getBoundaryNodeRange();
-    for (const auto & bnode : bnd_nodes)
+
+    if (!bnd_nodes.empty())
```

Comment:
OK - so I dug into this quite a bit more and learned several things. First, the boundary node range, as managed by `MooseMesh` works fine on DistributedMesh. We use the BoundaryInfo object in libMesh to get this information and it works properly only returning parts of the mesh that are accessible on each processor. From the docs:
``` 
 /**
   * Creates a list of nodes and ids for those nodes.
   *
   * On a ReplicatedMesh this will include all nodes; on a
   * DistributedMesh only semilocal nodes will be included.
   */
```

So, on a DistributedMesh, some processors may enter this branch, while others might not. I have verified that none of the contents of this branch contain and parallel communication.